flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClient.java

```diff
@@ -119,7 +145,66 @@ private void writeAccumulatedBuffers(
     private void writeAccumulatedBuffer(
             TieredStorageSubpartitionId subpartitionId, Buffer accumulatedBuffer)
             throws IOException {
-        // TODO, Try to write the accumulated buffer to the appropriate tier. After the tier is
-        // decided, then write the accumulated buffer to the tier.
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        int tierIndex = chooseStorageTierIfNeeded(subpartitionId);
+
+        Buffer compressedBuffer = compressBufferIfPossible(accumulatedBuffer);
+        updateStatistics(compressedBuffer);
+        boolean isLastBufferInSegment =
+                tierProducerAgents.get(tierIndex).write(subpartitionId, compressedBuffer);
+        if (isLastBufferInSegment) {
+            isSubpartitionSegmentFinished[subpartitionIndex] = true;
+        }
+    }
+
+    private int chooseStorageTierIfNeeded(TieredStorageSubpartitionId subpartitionId)
+            throws IOException {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (isSubpartitionSegmentFinished[subpartitionIndex]) {
+            // If the current segment is finished, this subpartition need choose a new storage tier.
+            internalChooseStorageTier(subpartitionId);
+            isSubpartitionSegmentFinished[subpartitionIndex] = false;
+        }
+        return currentSubpartitionTierIndex[subpartitionIndex];
+    }
+
+    private void internalChooseStorageTier(TieredStorageSubpartitionId subpartitionId)
+            throws IOException {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        int segmentIndex = currentSubpartitionSegmentId[subpartitionIndex];
+        int nextSegmentIndex = segmentIndex + 1;
+
+        for (int tierIndex = 0; tierIndex < tierProducerAgents.size(); ++tierIndex) {
+            if (tierProducerAgents
+                    .get(tierIndex)
```

Comment:
This can be simplified if we change `int[] currentSubpartitionTierIndex` to `TierProducerAgent currentSubpartitionTier`. We won't need to loop over the integers then get it from the list.

Reply:
Fixed. Replace it by `currentSubpartitionTierAgent`.